### PR TITLE
[percep] Added second bearing to Obstacle LCM

### PR DESCRIPTION
**Pull request by vidyasilai**
Added a right bearing variable to the Obstacle LCM struct, in order to let nav decide what clear path to take. Also added left and right bearing variables to the PCL class and obstacle_return classes.

## Compile Test
- [x] Offline Debug: ` ./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=true `
  - [x] No compiler errors or warnings
- [x] Offline Silent: ` ./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=false `
  - [x] No compiler errors or warnings
- [x] ZED Debug: ` ./jarvis build jetson/percep -o with_zed=true ar_detection=true obs_detection=true perception_debug=true `
  - [x] No compiler errors, only ZED warnings
- [x] ZED Silent: ` ./jarvis build jetson/percep -o with_zed=true ar_detection=true obs_detection=true perception_debug=false `
   - [x] No compiler errors, only ZED warnings

## Offline Debug Test
- [x] `./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=true`
- [x] Run code against test1 image set located in perception-files repository
- Check for:
   - [x]  Left and right path are being calculated
   - [x]  Calculated bearing is reasonable given viewer output
   - [x]  AR Tag ID is 2
   - [x]  Obstacles with large width are detected
   - [x]  Path is projected around not through large obstacles
   - [x]  Short obstacles are detected
   - [x]  Steep inclines are detected as obstacles
- [x] Run code against test2 image set located in perception-files repository
- Check for:
   - [x]  Left and right path are being calculated
   - [x]  Calculated bearing is reasonable given viewer output
   - [x]  AR Tag ID is 6
   - [x]  2 AR Tags detected simultaneously
   - [x]  Obstacles with large width are detected
   - [x]  Path is projected around not through large obstacles
   - [x]  Short obstacles are detected
   - [x]  Steep inclines are detected as obstacles

## Offline Silent Test
- [x] `./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=false`
- [x] Run `time ./jarvis exec jetson/percep` against test_case1 image set
- [x] Run `time ./jarvis exec jetson/percep` against test_case2 image set
- [ ] User value is less than or equal to \<TODO Determine Value\>

## Unit Test
- [x] ` ./jarvis build jetson/percep -o with_zed=true ar_detection=true obs_detection=true perception_debug=true `
- [x] `./jarvis exec lcm_tools_echo Obstacle "/obstacle"`
- [x] `./jarvis exec lcm_tools_echo TargetList "/target_list"`
- Test environment contains:
  - [x] 1 large obstacle ~ over 40cm tall
  - [x] 2 small obstacles ~  25cm-30cm tall
  - [x] 1 post
  - [x] 1 hill
- Test for:
  - [x] Detected large obstacle 6m away
  - [x] Detected small obstacle 2m away
  - [x] Detected clear path when 1.2m gap between 1 large and 1 small obstacle
  - [x] Left and right path are being calculated
  - [x] Calculated bearing is reasonable given viewer output of path
  - [x] AR Tag ID is the correct ID
  - [x] Obstacles with large width are detected
  - [x] Short obstacles are detected
  - [ ] FPS values don't exceed \<TODO find this value in Issue #552\>
  - [x] Obstacle.lcm in /obstacle is populated correctly
  - [x] TargetList.lcm in /target_list is populated correctly

## Integration Test
- [x] Create new branch based off of main
- [x] Checkout new branch
- [x] Cherry-Pick all PRs that need testing onto new branch
- [x] Compile all Software Packages
- [x] Run all code simultaneously
- [x] Obstacle.lcm in /obstacle is populated correctly
- [x] TargetList.lcm in /target_list is populated correctly
- [x] Sensors are Active
  - [x] Stereo Camera (ZED)
  - [x] Inertial Measurement Unit (IMU)
  - [x] Global Positioning System (GPS)

### CODE REVIEW

**Comment by raytitan** [comment]
Please update [this ICD](https://docs.google.com/document/d/1WzsQUxGt2jDygyt_J5W_eQBTJtCMHHeMhDp42rwpi78/edit)

File: rover_msgs/Obstacle.lcm [hunk 1]
```diff
@@ -3,4 +3,5 @@ package rover_msgs;
 struct Obstacle {
 	double bearing; // from straight ahead
 	double distance; // from straight ahead
+	double rightBearing;
```
Comment (mmartin4972) [inline rover_msgs/Obstacle.lcm:6]: Could you move the rightBearing to be under the bearing?

File: rover_msgs/Obstacle.lcm [hunk 2]
```diff
@@ -3,4 +3,5 @@ package rover_msgs;
...
 }
```
Comment (mmartin4972) [inline rover_msgs/Obstacle.lcm:7]: Can you add a line at the end of this file?

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

File: rover_msgs/Obstacle.lcm (hunk 1 above)
Comment (vidyasilai) [reply rover_msgs/Obstacle.lcm:5]: Fixed

File: rover_msgs/Obstacle.lcm (hunk 2 above)
Comment (vidyasilai) [reply rover_msgs/Obstacle.lcm]: Fixed

**Comment by aswarner** [comment]
It looks like some of the tests were not complete - can you update on the status of those? @vidyasilai Great thorough testing, this is an awesome model of what our test documentation should look like

File: jetson/percep/main.cpp [hunk 3]
```diff
@@ -160,6 +161,7 @@ int main() {
     #endif
     
     #endif
+    cout << "bearing: " << obstacleMessage.bearing << " right bearing: " << obstacleMessage.rightBearing << endl;
```
Comment (aswarner) [inline jetson/percep/main.cpp:164]: is there a reason to leave this in? looks like a debug print statement

File: jetson/percep/pcl.cpp [hunk 4]
```diff
@@ -279,27 +279,28 @@ double PCL::FindClearPath(const std::vector<std::vector<int>> &interest_points,
     std::vector<int> obstacles; //index of the leftmost and rightmost obstacles in path
     
     //Check Center Path
-    if(CheckPath(interest_points, viewer, obstacles, compareLine(0,-HALF_ROVER), compareLine(0,HALF_ROVER))) {
+    if(CheckPath(interest_points, viewer, obstacles, compareLine(0,-HALF_ROVER), compareLine(0,HALF_ROVER))){
         std::cout << "CENTER PATH IS CLEAR!!!" << std::endl;
```
Comment (aswarner) [inline jetson/percep/pcl.cpp:283]: see above: are debug print statements necessary?

File: jetson/percep/main.cpp (hunk 3 above)
Comment (vidyasilai) [reply jetson/percep/main.cpp:164]: Deleted

File: jetson/percep/pcl.cpp (hunk 4 above)
Comment (vidyasilai) [reply jetson/percep/pcl.cpp:283]: As per what Matt told me, since I didn't add this particular debug print statement, I won't delete this specific one.

**Comment by raytitan** [comment]
@mmartin4972 This branch now has conflicts fyi

File: jetson/percep/artag_detector.cpp [hunk 5]
```diff
@@ -162,6 +162,7 @@ void TagDetector::updateDetectedTagInfo(rover_msgs::Target *arTags, pair<Tag, Ta
         if(tags.buffer[i] <= BUFFER_ITERATIONS){//send buffered tag until tag is found
             ++tags.buffer[i];
         } else {//if still no tag found, set all stats to -1
+
```
Comment (mmartin4972) [inline jetson/percep/artag_detector.cpp:165]: Nit: Could you remove the new line here?

File: jetson/percep/pcl.cpp [hunk 6]
```diff
@@ -310,39 +310,28 @@ double PCL::FindClearPath(const std::vector<std::vector<int>> &interest_points)
...
-
+        //Find Clear left path, set left bearing
+        leftBearing = getAngleOffCenter(10, 0, interest_points, obstacles);
+        
         //Reset global variables
         obstacles = {0, 0};
         obstacles.at(0) = centerObstacles.at(0);
         obstacles.at(1) = centerObstacles.at(1);
 
-        //Find clear right path
```
Comment (mmartin4972) [inline jetson/percep/pcl.cpp:335]: I believe that the lines you deleted here are Kevin's depth code. Could you add that back in?

File: jetson/percep/pcl.hpp [hunk 7]
```diff
@@ -89,8 +90,7 @@ class PCL {
             viewer -> close();
             viewer_original -> close();
         #endif
-        };
-
+};
```
Comment (mmartin4972) [inline jetson/percep/pcl.hpp:93]: Could you fix the indenting of this curly bracket? It's kind of misleading

**Review by mmartin4972 — CHANGES_REQUESTED** [review]
Could you please test all of the silent mode parts of the checklist, and rerun our code against the two offline tests to make sure everything is still looking alright? A lot has changed since you last did those tests. Also, can you closely watch for the distance an obstacle is away when you do the tests to make sure we still have that functionality?

**Comment by vidyasilai** [comment]
I re ran the tests after fixing up Matt's comments and everything looks fine when I run it but I do get a couple compiler warnings in camera.cpp regarding uninitialized variables that I'm not so sure about

File: jetson/percep/artag_detector.cpp [hunk 8]
```diff
@@ -176,4 +175,4 @@ void TagDetector::updateDetectedTagInfo(rover_msgs::Target *arTags, pair<Tag, Ta
         tags.buffer[i] = 0;
    }
   }
-}
+}
```
Comment (mmartin4972) [inline jetson/percep/artag_detector.cpp:178]: Could you add a line to the end of this file?

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

File: jetson/percep/artag_detector.cpp [hunk 9]
```diff
@@ -74,10 +74,10 @@ pair<Tag, Tag> TagDetector::findARTags(Mat &src, Mat &depth_src, Mat &rgb) {  //
...
     // Find tags
     cv::aruco::detectMarkers(rgb, alvarDict, corners, ids, alvarParams);
-#if AR_RECORD
-cv::aruco::drawDetectedMarkers(rgb, corners, ids);
-#endif
-#if PERCEPTION_DEBUG
+    #if AR_RECORD
+    cv::aruco::drawDetectedMarkers(rgb, corners, ids);
+    #endif
+    #if PERCEPTION_DEBUG
```
Comment (mmartin4972) [inline jetson/percep/artag_detector.cpp:80]: Could you add a line in between these if statements?

File: jetson/percep/main.cpp [hunk 10]
```diff
@@ -143,12 +142,12 @@ int main() {
 
         //Run Obstacle Detection
         pointcloud.pcl_obstacle_detection();  
-        obstacle_return obstacle_detection (pointcloud.bearing, pointcloud.distance);
+        obstacle_return obstacle_detection (pointcloud.leftBearing, pointcloud.rightBearing, pointcloud.distance);
```
Comment (mmartin4972) [inline jetson/percep/main.cpp:145]: Could you rename obstacle_detection to be obstacleOutput? The current name is misleading

File: jetson/percep/pcl.cpp [hunk 11]
```diff
@@ -300,8 +300,8 @@ double PCL::getAngleOffCenter(int buffer, int direction, const std::vector<std::
 }
 
 /* --- Find Clear Path --- */
-//Returns the angle to a clear path
-double PCL::FindClearPath(const std::vector<std::vector<int>> &interest_points) {
+// Calculates left and right bearings
+void PCL::FindClearPath(const std::vector<std::vector<int>> &interest_points) {
```
Comment (mmartin4972) [inline jetson/percep/pcl.cpp:304]: Get rid of the extra spaces as the end of line 304

File: jetson/percep/pcl.cpp [hunk 12]
```diff
@@ -310,39 +310,38 @@ double PCL::FindClearPath(const std::vector<std::vector<int>> &interest_points)
...
         //Initialize base cases outside of scope
         vector<int> centerObstacles = {obstacles.at(0), obstacles.at(1)};
 
-        //Find Clear left path
-        double leftAngle = getAngleOffCenter(10, 0, interest_points, obstacles);
+        //Find clear left path, set left bearing
+        leftBearing = getAngleOffCenter(10, 0, interest_points, obstacles);
         leftDistance = distance; 
-
+
```
Comment (mmartin4972) [inline jetson/percep/pcl.cpp:331]: Get rid of the whitespace on line 331

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

**Review by mmartin4972 — APPROVED** [review]

**Review by michmli — APPROVED** [review]

File: jetson/percep/main.cpp [hunk 13]
```diff
@@ -143,23 +142,25 @@ int main() {
...
+            lastObstacle = obstacleOutput;
 
-        obstacleMessage.distance = lastObstacle.distance; //update LCM distance field
-        obstacleMessage.bearing = lastObstacle.bearing; //update LCM bearing field
+        //Update LCM 
+        obstacleMessage.bearing = lastObstacle.leftBearing; // Update LCM bearing field
+        obstacleMessage.rightBearing = lastObstacle.rightBearing;
+        obstacleMessage.distance = lastObstacle.distance; // Update LCM distance field
         #if PERCEPTION_DEBUG
             cout << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!Path Sent: " << obstacleMessage.bearing << "\n";
```
Comment (raytitan) [inline jetson/percep/main.cpp:165]: Is there a better way than this string of !!! lol

**Review by raytitan — APPROVED** [review]
Looks good! Only have one nit.

I'm happy to see comments all across the code as well

**Review by michmli — APPROVED** [review]

File: jetson/percep/main.cpp (hunk 13 above)
Comment (mmartin4972) [reply jetson/percep/main.cpp:165]: We're going to clean this up in a different PR

**Review by mmartin4972 — APPROVED** [review]

**Review by niwhsa9 — APPROVED** [review]

**Comment by vidyasilai** [comment]
Resolves issues #526 and #419. Adds a second bearing that is sent to nav for them to decide which clear path to take.

**Pull request merged**